src/python/grpcio_tests/tests_aio/unit/client_unary_stream_interceptor_test.py

```diff
@@ -0,0 +1,404 @@
+                self.assertEqual(await call.code(), grpc.StatusCode.OK)
+                self.assertEqual(await call.initial_metadata(), ())
+                self.assertEqual(await call.trailing_metadata(), ())
+                self.assertEqual(await call.details(), '')
+                self.assertEqual(await call.debug_error_string(), '')
+                self.assertEqual(call.cancel(), False)
+                self.assertEqual(call.cancelled(), False)
+                self.assertEqual(call.done(), True)
+
+                if interceptor_class == _UnaryStreamInterceptorWith_ResponseIterator:
+                    self.assertTrue(interceptor.response_iterator.response_cnt,
+                                    _NUM_STREAM_RESPONSES)
+
+                await channel.close()
+
+    async def test_add_done_callback(self):
+        for interceptor_class in (_UnaryStreamInterceptorEmpty,
+                                  _UnaryStreamInterceptorWith_ResponseIterator):
+
+            with self.subTest(name=interceptor_class):
+                interceptor = interceptor_class()
+
+                request = messages_pb2.StreamingOutputCallRequest()
+                for _ in range(_NUM_STREAM_RESPONSES):
+                    request.response_parameters.append(
+                        messages_pb2.ResponseParameters(
+                            size=_RESPONSE_PAYLOAD_SIZE))
+
+                channel = aio.insecure_channel(self._server_target,
+                                               interceptors=[interceptor])
+                stub = test_pb2_grpc.TestServiceStub(channel)
+                call = stub.StreamingOutputCall(request)
+
+                validation = _inject_callbacks(call)
+
+                async for response in call:
+                    pass
+
+                await validation
+
+                await channel.close()
+
+    async def test_add_done_callback_after_connection(self):
+        for interceptor_class in (_UnaryStreamInterceptorEmpty,
+                                  _UnaryStreamInterceptorWith_ResponseIterator):
+
+            with self.subTest(name=interceptor_class):
+                interceptor = interceptor_class()
+
+                request = messages_pb2.StreamingOutputCallRequest()
+                for _ in range(_NUM_STREAM_RESPONSES):
+                    request.response_parameters.append(
+                        messages_pb2.ResponseParameters(
+                            size=_RESPONSE_PAYLOAD_SIZE))
+
+                channel = aio.insecure_channel(self._server_target,
+                                               interceptors=[interceptor])
+                stub = test_pb2_grpc.TestServiceStub(channel)
+                call = stub.StreamingOutputCall(request)
+
+                # This ensures that the callbacks will be registered
+                # with the intercepted call rather than saving in the
+                # pending state list.
+                await call.wait_for_connection()
```

Comment:
How would it affect the behavior of this test if we removed this line? It seems like adding callbacks should be possible both before and after the channel has connected.

Reply:
I should rename the test, What I'm trying to test here is that the registration of the callback works as it is expected when the interceptor task has finished [1].

We have another test, the previous one [2], that tests that the registration of the callback works when the interceptor task is not yet finished.

I would rename both tests to make this explicit, WDYT?

[1] https://github.com/grpc/grpc/pull/22713/files#diff-173e8bd5d2219ba109f86237680068a4R329
[2] https://github.com/grpc/grpc/pull/22713/files#diff-13bced45184e6a42f0040ac938a37d87R157